tests/backend/test_bosonic_backend.py

```diff
@@ -0,0 +1,566 @@
+        implemented."""
+        with pytest.raises(NotImplementedError):
+            prog = sf.Program(1)
+            with prog.context as q:
+                sf.ops.GKP(desc="complex") | q[0]
+
+            backend = bosonic.BosonicBackend()
+            backend.run_prog(prog)
+
+
+class TestBosonicUserSpecifiedState:
+    """Checks the Bosonic preparation method."""
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    def test_complex_weight(self, alpha):
+        r"""Checks that Bosonic creates a state with user-specifed weights, means
+        and covariances."""
+        dummy_backend = bosonic.BosonicBackend()
+        dummy_backend.begin_circuit(1)
+        # Get weights, means and covs for a cat state
+        weights, means, covs = dummy_backend.prepare_cat(alpha, 0, 0, "complex", 0)
+
+        # Initiate the state, but use Bosonic method
+        backend = bosonic.BosonicBackend()
+        prog = sf.Program(1)
+        with prog.context as q1:
+            sf.ops.Bosonic(weights, means, covs) | q1[0]
+        backend.run_prog(prog)
+        state = backend.state()
+
+        # Prepare using the Catstate method
+        backend2 = bosonic.BosonicBackend()
+        prog2 = sf.Program(1)
+        with prog2.context as q2:
+            sf.ops.Catstate(alpha) | q2[0]
+        backend2.run_prog(prog2)
+        state2 = backend2.state()
+
+        # Check the two states are equal
+        assert state.__eq__(state2)
+
+
+class TestBosonicPrograms:
+    """Tests that small programs run and return the correct output."""
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_init_circuit(self, alpha, r):
+        """Checks init_circuit only prepares non-Gaussian states and vacuum."""
+        prog = sf.Program(3)
+        with prog.context as q:
+            sf.ops.Catstate(alpha) | q[0]
+            # this line should be ignored since it is a Gaussian prep
+            # that would be picked up in run_prog, but not init_circuit
+            sf.ops.Squeezed(r) | q[1]
+        backend = bosonic.BosonicBackend()
+        backend.init_circuit(prog)
+        state = backend.state()
+
+        # Check that second and third mode are still in vacuum
+        weights, means, covs = state.reduced_bosonic([1, 2])
+
+        mean_compare = np.zeros(4)
+        means_compare = np.tile(mean_compare, (4, 1))
+        assert np.allclose(means, means_compare)
+
+        cov_compare = np.diag([1, 1, 1, 1])
+        covs_compare = np.tile(cov_compare * sf.hbar / 2, (4, 1, 1))
+        assert np.allclose(covs, covs_compare)
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_different_preparations(self, alpha, r):
+        """Runs a program with non-Gaussian and Gaussian preparations."""
+        prog = sf.Program(3)
+        with prog.context as q:
+            sf.ops.Catstate(alpha) | q[0]
+            sf.ops.Squeezed(r) | q[1]
```

Comment:
Added new lines, recommend doing it at other places too.
```suggestion
        prog = sf.Program(3)

        with prog.context as q:
            sf.ops.Catstate(alpha) | q[0]
            sf.ops.Squeezed(r) | q[1]

```